c++/src/kj/async.c++

```diff
@@ -337,7 +329,24 @@ private:
   Own<_::PromiseNode> node;
 };
 
+TaskSet::~TaskSet() noexcept(false) {
+  // We need to make sure that we destroy the items in a loop to prevent any issues with stack
+  // overflow. We could use tasks being the null pointer as a signal of shutting down but that
+  // complicates pop() not working since tasks is required to point to the head of the list for that
+  // method to work.
+  shuttingDown = true;
+  while (tasks != nullptr) {
+    auto removed = KJ_REQUIRE_NONNULL(tasks)->pop();
+  }
+}
+
 void TaskSet::add(Promise<void>&& promise) {
+  if (shuttingDown) {
```

Comment:
Minor behavioural change since last review but since I was adding `shuttingDown` anyway so that `onEmpty` behaviour was correct.

Reply:
This change isn't clearly correct, because it means the promise will be destroyed synchronously, whereas previously it was destroyed asynchronously. What if the destruction takes a lock, but the caller already holds that lock? I would be more comfortable leaving it the way it was.

Reply:
Added a comment so that a future reader might think twice about this optimization as well.